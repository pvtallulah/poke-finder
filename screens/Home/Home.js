import React, {useState, useEffect, useRef} from 'react';
import {
  SafeAreaView,
  StyleSheet,
  View,
  StatusBar,
  Keyboard,
  Text,
} from 'react-native';

import {Colors} from 'react-native/Libraries/NewAppScreen';

import Alert from '../../components/Alert';
import Loader from '../../components/Loader';
import Textbox from '../../components/Textbox';
import Autocomplete from '../../components/Autocomplete';

import PokemonInfo from './components/PokemonInfo';
import PokemonMovements from './components/PokemonMovements';
import PokemonAbilities from './components/PokemonAbilities';
import PokemonSpecies from './components/PokemonSpecies';
import client from '../../api/';
import Row from '../../components/Row';

const Home = ({navigation}) => {
  const textboxRef = useRef(null);
  const [loading, setLoading] = useState(false);
  const [selectedPokemonName, setSelectedPokemonName] = useState('');
  const [selectedPokemon, setSelectedPokemon] = useState({});
  const [pokemonData, setPokemonData] = useState({});
  const [query, setQuery] = useState('');
  const [showAutocomplete, setShowAutocomplete] = useState(false);
  const [showAlert, setShowAlert] = useState(false);
  const [alertText, setAlertText] = useState('');
  const [pokemonList, setPokemonList] = useState([]);
  const initData = async () => {
    const pokemonCount = await getPokemonCount();
    const allPokemons = await getAllPokemons(pokemonCount);
    setPokemonList(allPokemons.results);
  };

  const getPokemonCount = async () => {
    try {
      setLoading(true);
      const countData = await client.GET('/pokemon?count&limit=1');
      return countData.count;
    } catch (error) {
      setAlertText(error.message);
      setShowAlert(true);
      return -1;
    } finally {
      setLoading(false);
    }
  };

  const getAllPokemons = async (pokemonCount) => {
    try {
      const pokemonList = await client.GET(`/pokemon?limit= ${pokemonCount}`);
      return pokemonList;
    } catch (error) {
      setAlertText(error.message);
      setShowAlert(true);
      return [];
    }
  };

  const pokemonSelected = async (pokemon) => {
    console.log('pokemon: ', pokemon);
    setSelectedPokemonName(pokemon.name);
    setShowAutocomplete(false);
    setSelectedPokemon(pokemon);
    const pokemonId = pokemon.url.split('/')[6];
    try {
      setLoading(true);
      const pokemonDataRes = await client.GET(`/pokemon/${pokemonId}/`);
      setPokemonData(pokemonDataRes);
    } catch (error) {
      setAlertText(error.message);
      setShowAlert(true);
    } finally {
      setLoading(false);
    }
  };

  const clearText = () => {
    setQuery('');
    setShowAutocomplete(false);
    setSelectedPokemon({});
    setSelectedPokemonName('');
    setPokemonData({});
    textboxRef.current.clear();
    Keyboard.dismiss();
  };

  const textboxFocus = () => {
    setShowAutocomplete(true);
  };

  useEffect(() => {
    initData();
  }, []);

  return (
    <>
      <StatusBar barStyle="dark-content" />
      <Alert
        message={alertText}
        setIsAlertVisible={setShowAlert}
        isAlertVisible={showAlert}
      />
      <Loader isLoaderVisible={loading} />
      <SafeAreaView>
        <View style={styles.scrollView}>
          <View style={styles.sectionContainer}>
            <Textbox
              onChange={(val) => setQuery(val)}
              forwardedRef={textboxRef}
              onFocus={textboxFocus}
              value={selectedPokemonName}
              setValue={setSelectedPokemonName}
              clearText={clearText}
            />
          </View>
          <View style={styles.sectionContainer}>
            <Autocomplete
              showList={showAutocomplete}
              data={pokemonList}
              query={query}
              noResultText="No se encontraron Pokemones."
              selected={pokemonSelected}
            />
          </View>
        </View>
        {Object.keys(pokemonData).length > 0 && (
          <>
            <View style={[styles.sectionContainer, styles.verticalSpace]}>
              <PokemonInfo
                header={pokemonData.name}
                weight={pokemonData.weight}
                height={pokemonData.height}
                baseExperience={pokemonData.base_experience}
              />
            </View>
            <View style={styles.twoCards}>
              <View style={[styles.sectionContainer, styles.verticalSpace]}>
                <PokemonMovements movements={pokemonData.moves} />
              </View>
              <View style={[styles.sectionContainer, styles.verticalSpace]}>
                <PokemonAbilities abilities={pokemonData.abilities} />
              </View>
            </View>
            <View style={[styles.sectionContainer, styles.verticalSpace]}>
              <PokemonSpecies id={pokemonData.id} />
            </View>
          </>
        )}
      </SafeAreaView>
    </>
  );
};

const styles = StyleSheet.create({
  scrollView: {
    backgroundColor: Colors.lighter,
  },
  engine: {
    position: 'absolute',
    right: 0,
  },
  body: {
    backgroundColor: Colors.white,
  },
  sectionContainer: {
    paddingHorizontal: 12,
  },
  verticalSpace: {
    marginTop: 15,
  },
  twoCards: {
    flexDirection: 'row',
  },
});

export default Home;
