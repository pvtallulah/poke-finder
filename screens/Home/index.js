import 'react-native-gesture-handler';
import React from 'react';
import {createStackNavigator} from '@react-navigation/stack';
import HomeScreen from './Home';
const HomeStack = createStackNavigator();

export default () => {
  return (
    <HomeStack.Navigator
      screenOptions={{
        headerShown: true,
      }}>
      <HomeStack.Screen
        initialRouteName="Home"
        name="Inicio"
        component={HomeScreen}
      />
    </HomeStack.Navigator>
  );
};
