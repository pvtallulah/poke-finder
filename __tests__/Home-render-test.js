import React from 'react';
import 'react-native';
import {NavigationContainer} from '@react-navigation/native';
import Home from '../screens/Home';

import renderer, {act, cleanup} from 'react-test-renderer';
jest.useFakeTimers();

test('renders correctly', async () => {
  afterEach(cleanup); //Unmounts React trees that were mounted with render to prevent memory leak.
  const tree = renderer
    .create(
      <NavigationContainer>
        <Home />
      </NavigationContainer>,
    )
    .toJSON();
  await act(() => {});
  expect(tree).toMatchSnapshot();
});

// jest.mock('../api');
// describe('Testing react navigation', () => {
//   test('page contains one textbox to find pokemons', async () => {
//     const component = (
//       <NavigationContainer>
//         <Home />
//       </NavigationContainer>
//     );

//     const {findByPlaceholderText} = render(component);

//     const textbox = await findByPlaceholderText(
//       'Ingrese por lo menos 3 caracteres',
//     );

//     expect(textbox).toBeTruthy();
//   });
// });
