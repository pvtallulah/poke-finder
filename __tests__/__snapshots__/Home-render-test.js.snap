// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "zIndex": 1,
          },
          false,
        ]
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        onLayout={[Function]}
        pointerEvents="box-none"
        style={null}
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "borderBottomColor": "rgb(216, 216, 216)",
                "flex": 1,
                "shadowColor": "rgb(216, 216, 216)",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.85,
                "shadowRadius": 0,
              }
            }
          />
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "maxHeight": undefined,
              "minHeight": undefined,
              "opacity": undefined,
              "transform": undefined,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 20,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "marginHorizontal": 16,
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityRole="header"
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 17,
                    "fontWeight": "600",
                  }
                }
              >
                Inicio
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "opacity": 1,
            }
          }
        />
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onGestureEnd={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              collapsable={false}
              forwardedRef={[Function]}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <Modal
                      animationType="slide"
                      hardwareAccelerated={false}
                      transparent={true}
                      visible={false}
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "rgba(0,0,0,0.5)",
                              "flex": 1,
                              "justifyContent": "center",
                              "paddingHorizontal": "10%",
                            },
                          ]
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "white",
                              "borderRadius": 0,
                              "maxWidth": "90%",
                              "minWidth": "90%",
                              "padding": 35,
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "width": "100%",
                              }
                            }
                          >
                            <Text
                              style={
                                Object {
                                  "fontSize": 25,
                                  "fontWeight": "bold",
                                  "marginBottom": 15,
                                  "textAlign": "left",
                                }
                              }
                            >
                              Pokefinder
                            </Text>
                          </View>
                          <View
                            style={
                              Object {
                                "marginBottom": 15,
                                "width": "100%",
                              }
                            }
                          >
                            <Text
                              style={
                                Object {
                                  "fontSize": 20,
                                  "textAlign": "left",
                                }
                              }
                            >
                              
                            </Text>
                          </View>
                          <View
                            style={
                              Object {
                                "flexDirection": "row",
                                "justifyContent": "space-evenly",
                                "width": "100%",
                              }
                            }
                          >
                            <View
                              accessible={true}
                              focusable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Object {
                                  "opacity": 1,
                                }
                              }
                            >
                              <Text
                                style={
                                  Object {
                                    "backgroundColor": "green",
                                    "borderWidth": 1,
                                    "color": "#fff",
                                    "fontSize": 18,
                                    "height": 35,
                                    "lineHeight": 35,
                                    "marginRight": 15,
                                    "textAlign": "center",
                                    "width": 150,
                                  }
                                }
                              >
                                Ok
                              </Text>
                            </View>
                          </View>
                        </View>
                      </View>
                    </Modal>
                    <Modal
                      animationType="slide"
                      hardwareAccelerated={false}
                      transparent={true}
                      visible={true}
                    >
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "backgroundColor": "rgba(0,0,0,0.2)",
                            "flex": 1,
                            "justifyContent": "center",
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "transparent",
                              "borderRadius": 20,
                              "padding": 35,
                              "width": "90%",
                            }
                          }
                        >
                          <ActivityIndicator
                            animating={true}
                            color="gray"
                            hidesWhenStopped={true}
                            size="large"
                          />
                        </View>
                      </View>
                    </Modal>
                    <RCTSafeAreaView
                      emulateUnlessSupported={true}
                    >
                      <View
                        style={
                          Object {
                            "backgroundColor": "#F3F3F3",
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "paddingHorizontal": 12,
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "backgroundColor": "#fff",
                                "borderWidth": 0.5,
                                "flexDirection": "row",
                                "height": 50,
                                "width": "100%",
                              }
                            }
                          >
                            <TextInput
                              allowFontScaling={true}
                              onChange={[Function]}
                              onFocus={[Function]}
                              placeholder="Ingrese por lo menos 3 caracteres"
                              rejectResponderTermination={true}
                              style={
                                Object {
                                  "fontSize": 18,
                                  "width": "94%",
                                }
                              }
                              underlineColorAndroid="transparent"
                              value=""
                            />
                            <RNGestureHandlerButton
                              collapsable={false}
                              onGestureEvent={[Function]}
                              onGestureHandlerEvent={[Function]}
                              onGestureHandlerStateChange={[Function]}
                              onHandlerStateChange={[Function]}
                              rippleColor={0}
                            >
                              <View
                                accessible={true}
                                style={
                                  Object {
                                    "alignContent": "center",
                                    "alignItems": "center",
                                    "alignSelf": "center",
                                    "borderRadius": 15,
                                    "borderWidth": 1,
                                    "flex": 1,
                                    "justifyContent": "center",
                                    "marginTop": 10,
                                    "maxHeight": 30,
                                    "maxWidth": 30,
                                    "minHeight": 30,
                                    "minWidth": 30,
                                    "opacity": 1,
                                  }
                                }
                              >
                                <Text
                                  style={
                                    Object {
                                      "fontSize": 20,
                                      "textAlign": "center",
                                    }
                                  }
                                >
                                  X
                                </Text>
                              </View>
                            </RNGestureHandlerButton>
                          </View>
                        </View>
                        <View
                          style={
                            Object {
                              "paddingHorizontal": 12,
                            }
                          }
                        />
                      </View>
                    </RCTSafeAreaView>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
